test/cpp/qps/scenario_runner.py

```diff
@@ -0,0 +1,138 @@
+# limitations under the License.
+"""
+Local QPS benchmark runner for the OSS Benchmark loadtest configurations.
+
+This tool will run a scenario locally, either already extracted from
+scenario_config_extractor, or extracted from a benchmark loadtest config. The
+driver, client, and server all in the same process. You can run the process
+under a custom runner using the --runner_cmd="<COMMAND>" flag, and with custom
+environment variables if needed.
+
+This example will run an optimized build of the loadtest under gdb
+
+GRPC_VERBOSITY=debug \
+    bazel run \
+    --config=opt \
+    --cxxopt="-gmlt" \
+    test/cpp/qps:scenario_runner -- \
+    --loadtest_file=/path/to/loadtest.config \
+    --runner_cmd="gdb --args"
+
+This builds the binary and runs:
+
+    gdb --args bazel-bin/.../scenario_runner -- \
+        --loadtest_config=/tmp/path/extracted_scenario_json.config
+
+        
+If you have already extracted the JSON scenario using scenario_config_exporter,
+you can replace `--loadtest_file=loadtest.yaml` with
+`--scenario_file=scenario.json`.
+
+
+Other --runner_cmd examples:
+    --runner_cmd="perf record -F 777 -o $(pwd)/perf.data -g --event=cpu-cycles",
+    --runner_cmd="perf stat record -o $(pwd)/perf.stat.data",
+"
+"""
+
+
+import os
+import subprocess
+import sys
+import tempfile
+
+from absl import app
+from absl import flags
+import yaml
+
+_LOADTEST_YAML = flags.DEFINE_string(
+    "loadtest_file", default=None, help="Path to the benchmark loadtest file"
+)
+_SCENARIO_JSON = flags.DEFINE_string(
+    "scenario_file", default=None, help="Path to a scenario JSON file"
+)
+_RUNNER_CMD = flags.DEFINE_string(
+    "runner_cmd",
+    default="",
+    help="Run the scearnio runner under a custom command (example: bazel ... --cmd='perf lock record -o $(pwd)/out')",
+)
+_RUN_FIRST = flags.DEFINE_bool(
+    "run_first",
+    default=False,
+    help="Only run the first scenario in the loadtest",
+)
+_RUN_ALL = flags.DEFINE_bool(
+    "run_all", default=False, help="Run all scenarios in the loadtest"
+)
+
+
+def run_command(filename):
+    cmd = [
+        os.path.join(
+            os.path.dirname(os.path.abspath(__file__)),
+            "scenario_runner_cc",
+        ),
+        "--loadtest_config",
+        filename,
+    ]
+    if _RUNNER_CMD.value:
+        cmd = _RUNNER_CMD.value.split(" ") + cmd
+    print(cmd)
+    subprocess.run(cmd, check=True)
+    if _RUN_FIRST.value:
+        print("Exiting due to --run_first")
+        sys.exit(0)
+
+
+def run_loadtests():
+    loadtests = []
+    with open(
+        os.path.join(
+            os.path.dirname(os.path.abspath(__file__)), _LOADTEST_YAML.value
+        )
+    ) as f:
+        loadtests = list(yaml.safe_load_all(f))
+    if len(loadtests) > 1 and not (_RUN_FIRST.value or _RUN_ALL.value):
+        print(
+            "The loadtest configuration contains more than one scenario. Please specify --run_first or --run_all.",
```

Comment:
Nit: The language being used here seems unclear. Each loadtest contains only one scenario. The "loadtest configuration file" may be a multipart yaml file containing multiple loadtests. So I would say "contains more than one loadtest".

It is also possible for a loadtest configuration file to contain multiple loadtests with the same scenario. This is the case for instance if they are generated with the `--runs_per_test` flag set.

https://github.com/grpc/grpc/blob/master/tools/run_tests/performance/README.md#generating-load-test-configurations

Reply:
The language is fixed, thanks.

And that's a good point regarding the repeated scenarios. We have room for future improvements. This script ignores `--runs_per_test` in the loadtest configuration, and if folks invoke scenario_runner with `--run_all`, it will happily run the same scenario again if present.